File: test-websocket.py
#!/usr/bin/env python3

"""
WebSocket Test Script for ERC-8004 Indexer (Python)

Usage:
    pip install websockets requests
    python test-websocket.py [username] [password] [api-url]

Example:
    python test-websocket.py admin admin123 https://api-8004-dev.fly.dev
"""

import asyncio
import json
import sys
import signal
from datetime import datetime
import requests
import websockets


class WebSocketTester:
    def __init__(self, api_url, username, password):
        self.api_url = api_url
        self.username = username
        self.password = password
        self.ws_url = api_url.replace('http://', 'ws://').replace('https://', 'wss://')
        self.event_count = 0
        self.connection_time = None
        self.running = True

    def login(self):
        """Login and get JWT token"""
        print(f"🔐 Logging in as '{self.username}'...")

        try:
            response = requests.post(
                f"{self.api_url}/login",
                json={"username": self.username, "password": self.password},
                timeout=10
            )
            response.raise_for_status()
            token = response.json()['token']
            print("✅ Login successful!\n")
            return token
        except requests.exceptions.RequestException as e:
            print(f"❌ Login error: {e}")
            sys.exit(1)

    async def connect_websocket(self, token):
        """Connect to WebSocket and handle events"""
        url = f"{self.ws_url}/ws?token={token}"
        print(f"🔌 Connecting to WebSocket...")
        print(f"   URL: {url}\n")

        try:
            async with websockets.connect(url) as websocket:
                self.connection_time = datetime.now()
                print("✅ WebSocket connected!")
                print("📊 Waiting for events...\n")
                print("Press Ctrl+C to disconnect\n")
                print("─" * 80)

                while self.running:
                    try:
                        message = await asyncio.wait_for(websocket.recv(), timeout=1.0)
                        self.handle_message(message)
                    except asyncio.TimeoutError:
                        continue
                    except websockets.exceptions.ConnectionClosed as e:
                        print(f"\n🔌 Connection closed: {e}")
                        break

        except websockets.exceptions.WebSocketException as e:
            print(f"❌ WebSocket error: {e}")
        except Exception as e:
            print(f"❌ Unexpected error: {e}")
        finally:
            self.print_summary()

    def handle_message(self, message):
        """Handle incoming WebSocket message"""
        self.event_count += 1
        timestamp = datetime.now().isoformat()

        try:
            event = json.loads(message)

            print(f"\n📦 Event #{self.event_count} received at {timestamp}")
            print("─" * 80)
            print(f"Contract:     {event.get('contract_address', 'N/A')}")
            print(f"Event Type:   {event.get('event_type', 'N/A')}")
            print(f"Block:        {event.get('block_number', 'N/A')}")
            print(f"TX Hash:      {event.get('transaction_hash', 'N/A')}")
            print(f"Log Index:    {event.get('log_index', 'N/A')}")

            if event.get('block_timestamp'):
                block_time = datetime.fromtimestamp(event['block_timestamp'])
                print(f"Timestamp:    {block_time.isoformat()}")

            if event.get('data'):
                print("\nEvent Data:")
                print(json.dumps(event['data'], indent=2))

            print("─" * 80)

        except json.JSONDecodeError as e:
            print(f"⚠️  Failed to parse event: {e}")
            print(f"Raw data: {message}")

    def print_summary(self):
        """Print connection summary"""
        if self.connection_time:
            duration = (datetime.now() - self.connection_time).total_seconds()
            print(f"\n📊 Connection Summary:")
            print(f"   Duration: {duration:.2f}s")
            print(f"   Events received: {self.event_count}")
            if self.event_count > 0:
                print(f"   Avg rate: {self.event_count / duration:.2f} events/sec")

    def handle_shutdown(self, signum, frame):
        """Handle graceful shutdown"""
        print("\n\n⏹️  Shutting down...")
        self.running = False


async def main():
    # Parse command line arguments
    username = sys.argv[1] if len(sys.argv) > 1 else 'admin'
    password = sys.argv[2] if len(sys.argv) > 2 else 'admin123'
    api_url = sys.argv[3] if len(sys.argv) > 3 else 'https://api-8004-dev.fly.dev'

    print("🔐 ERC-8004 WebSocket Test")
    print("========================\n")

    tester = WebSocketTester(api_url, username, password)

    # Setup signal handler for graceful shutdown
    signal.signal(signal.SIGINT, tester.handle_shutdown)

    # Login and connect
    token = tester.login()
    await tester.connect_websocket(token)


if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        print("\n👋 Goodbye!")
    except Exception as e:
        print(f"❌ Fatal error: {e}")
        sys.exit(1)

File: test-websocket.js
#!/usr/bin/env node

/**
 * WebSocket Test Script for ERC-8004 Indexer
 *
 * Usage:
 *   npm install ws node-fetch
 *   node test-websocket.js [username] [password] [api-url]
 *
 * Example:
 *   node test-websocket.js admin admin123 https://api-8004-dev.fly.dev
 */

const WebSocket = require('ws');
const fetch = require('node-fetch');

// Parse command line arguments
const username = process.argv[2] || 'admin';
const password = process.argv[3] || 'admin123';
const apiUrl = process.argv[4] || 'https://api-8004-dev.fly.dev';

// Convert HTTP(S) URL to WS(S) URL
const wsUrl = apiUrl.replace(/^http/, 'ws');

console.log('🔐 ERC-8004 WebSocket Test');
console.log('========================\n');

async function login() {
    console.log(`📡 Logging in as '${username}'...`);

    try {
        const response = await fetch(`${apiUrl}/login`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify({ username, password })
        });

        if (!response.ok) {
            const error = await response.text();
            throw new Error(`Login failed: ${response.status} - ${error}`);
        }

        const data = await response.json();
        console.log('✅ Login successful!\n');
        return data.token;
    } catch (error) {
        console.error('❌ Login error:', error.message);
        process.exit(1);
    }
}

function connectWebSocket(token) {
    const url = `${wsUrl}/ws?token=${token}`;
    console.log(`🔌 Connecting to WebSocket...`);
    console.log(`   URL: ${url}\n`);

    const ws = new WebSocket(url);
    let eventCount = 0;
    let connectionTime = null;

    ws.on('open', () => {
        connectionTime = Date.now();
        console.log('✅ WebSocket connected!');
        console.log('📊 Waiting for events...\n');
        console.log('Press Ctrl+C to disconnect\n');
        console.log('─'.repeat(80));
    });

    ws.on('message', (data) => {
        eventCount++;
        const timestamp = new Date().toISOString();

        try {
            const event = JSON.parse(data.toString());

            console.log(`\n📦 Event #${eventCount} received at ${timestamp}`);
            console.log('─'.repeat(80));
            console.log(`Contract:     ${event.contract_address}`);
            console.log(`Event Type:   ${event.event_type}`);
            console.log(`Block:        ${event.block_number}`);
            console.log(`TX Hash:      ${event.transaction_hash}`);
            console.log(`Log Index:    ${event.log_index}`);
            console.log(`Timestamp:    ${new Date(event.block_timestamp * 1000).toISOString()}`);

            if (event.data) {
                console.log('\nEvent Data:');
                console.log(JSON.stringify(event.data, null, 2));
            }

            console.log('─'.repeat(80));
        } catch (error) {
            console.error('⚠️  Failed to parse event:', error.message);
            console.log('Raw data:', data.toString());
        }
    });

    ws.on('error', (error) => {
        console.error('\n❌ WebSocket error:', error.message);
    });

    ws.on('close', (code, reason) => {
        const duration = connectionTime ? ((Date.now() - connectionTime) / 1000).toFixed(2) : 'N/A';
        console.log(`\n🔌 WebSocket closed`);
        console.log(`   Code: ${code}`);
        console.log(`   Reason: ${reason || 'No reason provided'}`);
        console.log(`   Duration: ${duration}s`);
        console.log(`   Events received: ${eventCount}`);
    });

    ws.on('ping', () => {
        console.log('🏓 Ping received from server');
    });

    ws.on('pong', () => {
        console.log('🏓 Pong received from server');
    });

    // Handle graceful shutdown
    process.on('SIGINT', () => {
        console.log('\n\n⏹️  Shutting down...');
        ws.close(1000, 'Client disconnecting');
        setTimeout(() => process.exit(0), 1000);
    });
}

// Main execution
(async () => {
    try {
        const token = await login();
        connectWebSocket(token);
    } catch (error) {
        console.error('❌ Fatal error:', error.message);
        process.exit(1);
    }
})();
